### PR TITLE
x64 new backend: most integer opcodes + baldrdash + a bunch of codegen fixes

### PR DESCRIPTION
**Pull request by bnjbvr**
I've implemented basic support for Baldrdash so as to be able to run the extensive Spidermonkey test suite, and then could get back to fix a few integer operators implemented before. Also implemented most of the integer opcodes required for wasm; I could start to run successfully a few large wasm integer-only programs with wasmtime too.

I plan on enabling limited testing of wast files on CI soon, to catch future regressions in codegen.

Sorry about the large PR!

### CODE REVIEW

File: cranelift/codegen/src/isa/x64/inst/args.rs [hunk 1]
```diff
@@ -509,6 +531,8 @@ pub enum ShiftKind {
     Left,
     RightZ,
     RightS,
+    RotateLeft,
+    RotateRight,
```
Comment (jlb6740) [inline cranelift/codegen/src/isa/x64/inst/args.rs:535]: Rotate in shift enum ... making the naming a little inconsistent from the first 3? Perhaps a more generic name (and comment update) for the Enum to show that Rotate is also housed. Also for consistency could prepend shift to Left, RightZ, and RightS, i.e. ShiftLeft, ShiftRightZ, ShiftRightS.

Comment (jlb6740) [inline cranelift/codegen/src/isa/x64/inst/args.rs:535]: Rotate in shift enum ... making the naming a little inconsistent from the first 3? Perhaps a more generic name (and comment update) for the Enum to show that Rotate is also housed. Also for consistency could prepend shift to Left, RightZ, and RightS, i.e. ShiftLeft, ShiftRightZ, ShiftRightS.

Comment (julian-seward1) [reply cranelift/codegen/src/isa/x64/inst/args.rs:535]: +1 for that suggestion.

File: cranelift/codegen/src/isa/x64/inst/mod.rs [hunk 2]
```diff
@@ -57,6 +57,9 @@ pub enum Inst {
         loc: SourceLoc,
     },
 
+    /// The high result of a (un)signed multiply: imul/mul using RAX:RDX.
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/inst/mod.rs:69]: Could you clarify the register usage in this comment a bit?  Something like: `RDX:RAX <- RAX * rhs`.

File: cranelift/codegen/src/isa/x64/inst/emit.rs [hunk 3]
```diff
@@ -1403,8 +1536,31 @@ pub(crate) fn emit(
...
             emit_std_reg_mem(sink, prefix, opcode, 2, *src, dst, rex);
         }
+
+        Inst::LoadExtName {
+            dst,
+            name,
+            offset,
+            srcloc,
+        } => {
+            // The full address can be encoded in the register, with a relocation.
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/inst/emit.rs:1552]: Add a comment giving the instruction this is emitting? `movq $imm64, %dst` or similar...

File: cranelift/codegen/src/isa/x64/lower.rs [hunk 4]
```diff
@@ -747,6 +747,29 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
...
+            let input_ty = ctx.input_ty(insn, 0);
+            let size = input_ty.bytes() as u8;
+
+            let lhs = input_to_reg(ctx, inputs[0]);
+            let rhs = input_to_reg_mem(ctx, inputs[1]);
+            let dst = output_to_reg(ctx, outputs[0]);
+
+            // Move lhs in %rax.
+            ctx.emit(Inst::gen_move(
+                Writable::from_reg(regs::rax()),
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/lower.rs:1167]: Hmm, `gen_move` has its destination on the left?  That feels inconsistent with the rest of the x64 stuff, which I thought had destinations as the rightmost argument.  But maybe that's because `gen_move` is from the applies-to-all-targets `trait Inst`, right?  I guess the use of `Writable` guarantees we're safe here.

File: cranelift/codegen/src/isa/x64/inst/args.rs [hunk 5]
```diff
@@ -289,6 +289,26 @@ impl ToString for AluRmiROpcode {
     }
 }
 
+#[derive(Clone, PartialEq)]
+pub enum ReadOnlyGprRmROpcode {
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/inst/args.rs:292]: `ReadOnlyGprRmR*` feels a bit cumbersome to me (and did when I looked at it some days ago).  Would you consider instead `UnaryRmR` (and consistent renaming for the derived-from-it names?)

File: cranelift/codegen/src/isa/x64/inst/mod.rs [hunk 6]
```diff
@@ -57,17 +66,24 @@ pub enum Inst {
...
+    /// Puts the result back into %rax if is_div, %rdx if !is_div, to mimic what the div
     /// instruction does.
     /// The generated code sequence is described in the emit's function match arm for this
     /// instruction.
+    ///
+    /// Note: %rdx is marked as modified by this instruction, to avoid an early clobber problem
+    /// with the temporary and divisor. Make sure to zero %rdx right before this instruction!
     CheckedDivOrRemSeq {
         is_div: bool,
         is_signed: bool,
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/inst/mod.rs:83]: Drive-by / pre-existing, but perhaps we could turn `is_div` and `is_signed` into their own (combined) enum -- SDiv, UDiv, SRem, URem?

File: cranelift/codegen/src/isa/x64/inst/mod.rs [hunk 7]
```diff
@@ -1243,10 +1389,11 @@ impl MachInst for Inst {
         Inst::jmp_known(BranchTarget::Label(label))
     }
 
-    fn gen_constant(to_reg: Writable<Reg>, value: u64, _: Type) -> SmallVec<[Self; 4]> {
+    fn gen_constant(to_reg: Writable<Reg>, value: u64, ty: Type) -> SmallVec<[Self; 4]> {
         let mut ret = SmallVec::new();
-        let is64 = value > 0xffff_ffff;
-        ret.push(Inst::imm_r(is64, value, to_reg));
+        debug_assert!(ty.is_int(), "float constants NYI");
+        let is_64 = ty == I64 && value > 0x7fffffff;
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/inst/mod.rs:1395]: Perhaps for a future refactor, but IMHO we should move from `is_64` bools to a purpose-built enum.

File: cranelift/codegen/src/isa/x64/lower.rs [hunk 8]
```diff
@@ -120,12 +120,51 @@ struct InsnOutput {
...
 
-fn input_to_reg<'a>(ctx: Ctx<'a>, spec: InsnInput) -> Reg {
+fn input_to_reg(ctx: Ctx, spec: InsnInput) -> Reg {
     let inputs = ctx.get_input(spec.insn, spec.input);
     ctx.use_input_reg(inputs);
     inputs.reg
 }
 
+enum ExtSpec {
+    ZeroExtend32,
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/lower.rs:131]: nit-level, at best, but .. (eg) `*Extend32` doesn't make it obvious to the casual reader whether the 32 is the source or dest width.  Would you consider adding `To` in front of the number?  eg `*ExtendTo32` ?

File: cranelift/codegen/src/isa/x64/inst/regs.rs [hunk 9]
```diff
@@ -33,99 +33,108 @@ fn gpr(enc: u8, index: u8) -> Reg {
...
+}
+
+pub(crate) fn r15() -> Reg {
+    // r15 is put aside since this is the pinned register.
+    gpr(ENC_R15, 29)
+}
+
+/// The pinned register on this architecture.
+/// It must be the same as Spidermonkey's HeapReg, as found in this file.
+/// https://searchfox.org/mozilla-central/source/js/src/jit/x64/Assembler-x64.h#99
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/inst/regs.rs:82]: I know this is just following the aarch64 backend (so it's OK for now!), but I'm wondering if we should make the pinned register a flags option, and then set it explicitly in Baldrdash, to avoid the hardcoded reference to SpiderMonkey here. Perhaps for a future refactor?

File: cranelift/codegen/src/isa/x64/lower.rs [hunk 10]
```diff
@@ -390,6 +390,238 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
             ));
         }
 
+        Opcode::Popcnt => {
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/lower.rs:394]: I'm sure this works correctly, but it concerns me in two ways:

* The text here is long and it's (almost) exactly a 6-way unrolled loop in the 64-bit case.  Would it be possible to do this by having a loop here (in lower.rs, not the generated code) which reads a table of constants?  In the style of https://sourceware.org/git/?p=valgrind.git;a=blob;f=VEX/priv/guest_amd64_toIR.c;h=fadf47d41d6ba6f63a5ca3800409b11da94c5f7e;hb=HEAD#l4948

* I see an integer multiply at the end.  In the sequences I'm familiar with, that doesn't appear.  It's probably a 3-cycle latency, compared with the 1-cycle latency and parallelisable over #-available-basic-ALUs for a sequence using only and/or/add/sub/shifts.

Given that the current goal is to get correct x64 code generation landed, I'd be happy for these to be looked at in a followup.  They don't need to block landing.

File: cranelift/codegen/src/isa/x64/inst/emit.rs (hunk 3 above)
Comment (julian-seward1) [reply cranelift/codegen/src/isa/x64/inst/emit.rs:1552]: +1 for that; please add comments always, indicating what insns are being emitted.  Inferring from the opcodes is nearly impossible.

Comment (julian-seward1) [reply cranelift/codegen/src/isa/x64/inst/emit.rs:1552]: Also .. is there a typo in the commit message?  "movaps" isn't the insn you mean AFAIK; that's an aligned SSE load (move-aligned-packed-singleprecision).  Maybe you mean "movabsq" ?

File: cranelift/codegen/src/isa/x64/abi.rs [hunk 11]
```diff
@@ -92,7 +92,11 @@ fn in_vec_reg(ty: types::Type) -> bool {
     }
 }
 
-fn get_intreg_for_arg_systemv(idx: usize) -> Option<Reg> {
+fn get_intreg_for_arg_systemv(call_conv: &CallConv, idx: usize) -> Option<Reg> {
+    assert!(match call_conv {
+        CallConv::SystemV | CallConv::BaldrdashSystemV => true,
+        _ => false,
```
[the remark this answers is not in the document]
Comment (julian-seward1) [reply cranelift/codegen/src/isa/x64/abi.rs:98]: If it's an outright bug that should never happen, then +1 for panicing rather than continuing (returning `false` or whatever).  I'm not convinced that's the case here, though (can't infer)

File: cranelift/codegen/src/isa/x64/inst/regs.rs [hunk 12]
```diff
@@ -33,99 +33,108 @@ fn gpr(enc: u8, index: u8) -> Reg {
 }
 
 pub(crate) fn r12() -> Reg {
-    gpr(ENC_R12, 0)
+    gpr(ENC_R12, 16)
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/inst/regs.rs:36]: I just want to check here that these changes to the register numbering don't change the order at which they appear in the final `RealRegUniverse`.  (To be specific: you are not reordering registers within each reg-class, in the universe).  Is that true?  I ask because the ordering of regs within each reg class is relevant -- at least for RA/BT -- to whether caller- or callee-saved registers are used in preference, and that has an effect on the quality of the generated allocation.

File: cranelift/codegen/src/isa/x64/inst/regs.rs (hunk 9 above)
Comment (julian-seward1) [reply cranelift/codegen/src/isa/x64/inst/regs.rs:82]: +1; losing `r15` in configurations where that isn't necessary would be unfortunate, given that there are only circa 12 int regs available for allocation anyway.  And in the 32-bit x86 case (if we ever do that) it would be quite serious.

File: cranelift/codegen/src/isa/x64/lower.rs [hunk 13]
```diff
@@ -390,6 +390,238 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
...
+                    is_64,
+                    AluRmiROpcode::Mul,
+                    RegMemImm::reg(cst.to_reg()),
+                    dst,
+                ));
+
+                // shr $56, dst
+                ctx.emit(Inst::shift_r(is_64, ShiftKind::RightZ, Some(56), dst));
+            } else {
+                debug_assert_eq!(ty, I32);
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/lower.rs:522]: Please can this be upgraded to a release assert, or at least somehow be checked even in release builds?  If for whatever reason `ty` of `I16` or `I8` ever gets here, we will silently generate wrong code and it could be hard to track down.

File: cranelift/codegen/src/isa/x64/inst/mod.rs (hunk 6 above)
Comment (julian-seward1) [reply cranelift/codegen/src/isa/x64/inst/mod.rs:83]: Also .. for "Make sure to zero %rdx right before this instruction!" please add a line explaining why this is necessary and what will (or might) happen if you don't.

File: cranelift/codegen/src/isa/x64/abi.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (bjorn3) [reply cranelift/codegen/src/isa/x64/abi.rs:98]: `CallConv::Fast` and `CallConv::Cold` can also reach this I think.

File: cranelift/codegen/src/isa/x64/inst/args.rs [hunk 14]
```diff
@@ -289,6 +288,29 @@ impl ToString for AluRmiROpcode {
...
+impl fmt::Debug for ReadOnlyGprRmROpcode {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ReadOnlyGprRmROpcode::Bsr => write!(fmt, "bsr"),
+            ReadOnlyGprRmROpcode::Bsf => write!(fmt, "bsf"),
+        }
+    }
+}
+
+impl ToString for ReadOnlyGprRmROpcode {
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/x64/inst/args.rs:308]: If you implement `Display` instead, you get `ToString` for free.

File: cranelift/codegen/src/isa/x64/inst/regs.rs (hunk 9 above)
Comment (bjorn3) [reply cranelift/codegen/src/isa/x64/inst/regs.rs:82]: It is already conditionally used as pinned register: https://github.com/bytecodealliance/wasmtime/pull/1995/files#diff-d4cd987e797b464cfda540d1f41fcfb3R200-R210

File: cranelift/codegen/src/isa/x64/lower.rs (hunk 4 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/lower.rs:1167]: Indeed, `gen_move` is on the target-independent trait. Maybe we can change this at some point, if the aarch64 backend agrees with the x64 convention?

File: cranelift/codegen/src/isa/x64/inst/mod.rs (hunk 7 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/inst/mod.rs:1395]: Yes! Recent PRs with `OperandSize` enum have inspired me... Added to the backlog!

File: cranelift/codegen/src/isa/x64/inst/regs.rs (hunk 9 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/inst/regs.rs:82]: +1 to what @bjorn3 said, its usage is opt-in, disabled by default.

File: cranelift/codegen/src/isa/x64/lower.rs (hunk 10 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/lower.rs:394]: I've mostly ported what the old backend did, and I'm happy to revisit it later when this has landed. Thanks for the comments!

File: cranelift/codegen/src/isa/x64/inst/regs.rs (hunk 12 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/inst/regs.rs:36]: It's a long story, but... yes, this does put R15 at the end of the allocatable list.

The reason is the following: regalloc has the concept of overall allocatable registers. So if in certain modes (i.e. when the use_pinned_reg flag is true), we want to mark one register as allocatable or not, then it must be at the edge of the "allocatable" list, so at the end of its register class (otherwise its register index changes when the pinned reg is in use).

To keep BT happy with the order of caller or callee-saved registers, I guess we could have a different register index for r15, according to whether it's pinned or not. I think it might cause more churn for other registers (namely, those which will be placed after r15). So maybe this justifies a completely different `RegUniverse` of its own? If that's ok, i'd like to defer this. We may need having different RegUniverse for respecting the win64 calling convention.

Comment (julian-seward1) [reply cranelift/codegen/src/isa/x64/inst/regs.rs:36]: > If that's ok, i'd like to defer this.

Oh, definitely defer.  Let's get this landed and running first, then we can come back and optimise later.  (Sorry .. I should have made this clearer in the comment)

File: cranelift/codegen/src/isa/x64/lower.rs (hunk 4 above)
Comment (julian-seward1) [reply cranelift/codegen/src/isa/x64/lower.rs:1222]: Well, aarch64 has destinations on the left.  So it's never going to agree with x64 (the AT&T conventions, at least.).
